From 850c89022d9f8f551c1e9292646bf127a05ca871 Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Thu, 28 Aug 2014 14:22:56 -0400
Subject: [cpufreq] remove sysfs files for CPUs which failed to come back after resume

Message-id: <1409235879-1283-83-git-send-email-prarit@redhat.com>
Patchwork-id: 89417
O-Subject: [RHEL7.1 PATCH BZ 1134639 082/185] cpufreq: remove sysfs files for CPUs which failed to come back after resume
Bugzilla: 1134369
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Mark Langsdorf <mlangsdo@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1134639

commit 42f921a6f10c6c2079b093a115eb7e3c3508357f
Author: Viresh Kumar <viresh.kumar@linaro.org>
Date:   Fri Dec 20 21:26:02 2013 +0530

    cpufreq: remove sysfs files for CPUs which failed to come back after resume

    There are cases where cpufreq_add_dev() may fail for some CPUs
    during system resume. With the current code we will still have
    sysfs cpufreq files for those CPUs and struct cpufreq_policy
    would be already freed for them. Hence any operation on those
    sysfs files would result in kernel warnings.

    Example of problems resulting from resume errors (from Bjorn Mork):

    WARNING: CPU: 0 PID: 6055 at fs/sysfs/file.c:343 sysfs_open_file+0x77/0x212()
    missing sysfs attribute operations for kobject: (null)
    Modules linked in: [stripped as irrelevant]
    CPU: 0 PID: 6055 Comm: grep Tainted: G      D      3.13.0-rc2 #153
    Hardware name: LENOVO 2776LEG/2776LEG, BIOS 6EET55WW (3.15 ) 12/19/2011
     0000000000000009 ffff8802327ebb78 ffffffff81380b0e 0000000000000006
     ffff8802327ebbc8 ffff8802327ebbb8 ffffffff81038635 0000000000000000
     ffffffff811823c7 ffff88021a19e688 ffff88021a19e688 ffff8802302f9310
    Call Trace:
     [<ffffffff81380b0e>] dump_stack+0x55/0x76
     [<ffffffff81038635>] warn_slowpath_common+0x7c/0x96
     [<ffffffff811823c7>] ? sysfs_open_file+0x77/0x212
     [<ffffffff810386e3>] warn_slowpath_fmt+0x41/0x43
     [<ffffffff81182dec>] ? sysfs_get_active+0x6b/0x82
     [<ffffffff81182382>] ? sysfs_open_file+0x32/0x212
     [<ffffffff811823c7>] sysfs_open_file+0x77/0x212
     [<ffffffff81182350>] ? sysfs_schedule_callback+0x1ac/0x1ac
     [<ffffffff81122562>] do_dentry_open+0x17c/0x257
     [<ffffffff8112267e>] finish_open+0x41/0x4f
     [<ffffffff81130225>] do_last+0x80c/0x9ba
     [<ffffffff8112dbbd>] ? inode_permission+0x40/0x42
     [<ffffffff81130606>] path_openat+0x233/0x4a1
     [<ffffffff81130b7e>] do_filp_open+0x35/0x85
     [<ffffffff8113b787>] ? __alloc_fd+0x172/0x184
     [<ffffffff811232ea>] do_sys_open+0x6b/0xfa
     [<ffffffff811233a7>] SyS_openat+0xf/0x11
     [<ffffffff8138c812>] system_call_fastpath+0x16/0x1b

    To fix this, remove those sysfs files or put the associated kobject
    in case of such errors. Also, to make it simple, remove the cpufreq
    sysfs links from all the CPUs (except for the policy->cpu) during
    suspend, as that operation won't result in a loss of sysfs file
    permissions and we can create those links during resume just fine.

    Fixes: 5302c3fb2e62 ("cpufreq: Perform light-weight init/teardown during suspend/resume")
    Reported-and-tested-by: Bjorn Mork <bjorn@mork.no>
    Signed-off-by: Viresh Kumar <viresh.kumar@linaro.org>
    Cc: 3.12+ <stable@vger.kernel.org> # 3.12+
    [rjw: Changelog]
    Signed-off-by: Rafael J. Wysocki <rafael.j.wysocki@intel.com>

Cc: Lenny Szubowicz <lszubowi@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/cpufreq/cpufreq.c b/drivers/cpufreq/cpufreq.c
index 8776fab..50414da 100644
--- a/drivers/cpufreq/cpufreq.c
+++ b/drivers/cpufreq/cpufreq.c
@@ -845,8 +845,7 @@ static void cpufreq_init_policy(struct cpufreq_policy *policy)
 
 #ifdef CONFIG_HOTPLUG_CPU
 static int cpufreq_add_policy_cpu(struct cpufreq_policy *policy,
-      unsigned int cpu, struct device *dev,
-      bool frozen)
+      unsigned int cpu, struct device *dev)
 {
  int ret = 0;
  unsigned long flags;
@@ -877,11 +876,7 @@ static int cpufreq_add_policy_cpu(struct cpufreq_policy *policy,
   }
  }
 
- /* Don't touch sysfs links during light-weight init */
- if (!frozen)
-  ret = sysfs_create_link(&dev->kobj, &policy->kobj, "cpufreq");
-
- return ret;
+ return sysfs_create_link(&dev->kobj, &policy->kobj, "cpufreq");
 }
 #endif
 
@@ -926,6 +921,27 @@ err_free_policy:
  return NULL;
 }
 
+static void cpufreq_policy_put_kobj(struct cpufreq_policy *policy)
+{
+ struct kobject *kobj;
+ struct completion *cmp;
+
+ down_read(&policy->rwsem);
+ kobj = &policy->kobj;
+ cmp = &policy->kobj_unregister;
+ up_read(&policy->rwsem);
+ kobject_put(kobj);
+
+ /*
+  * We need to make sure that the underlying kobj is
+  * actually not referenced anymore by anybody before we
+  * proceed with unloading.
+  */
+ pr_debug("waiting for dropping of refcount\n");
+ wait_for_completion(cmp);
+ pr_debug("wait complete\n");
+}
+
 static void cpufreq_policy_free(struct cpufreq_policy *policy)
 {
  free_cpumask_var(policy->related_cpus);
@@ -986,7 +1002,7 @@ static int __cpufreq_add_dev(struct device *dev, struct subsys_interface *sif,
  list_for_each_entry(tpolicy, &cpufreq_policy_list, policy_list) {
   if (cpumask_test_cpu(cpu, tpolicy->related_cpus)) {
    read_unlock_irqrestore(&cpufreq_driver_lock, flags);
-   ret = cpufreq_add_policy_cpu(tpolicy, cpu, dev, frozen);
+   ret = cpufreq_add_policy_cpu(tpolicy, cpu, dev);
    up_read(&cpufreq_rwsem);
    return ret;
   }
@@ -1095,7 +1111,10 @@ err_get_freq:
  if (cpufreq_driver->exit)
   cpufreq_driver->exit(policy);
 err_set_policy_cpu:
+ if (frozen)
+  cpufreq_policy_put_kobj(policy);
  cpufreq_policy_free(policy);
+
 nomem_out:
  up_read(&cpufreq_rwsem);
  return ret;
@@ -1116,7 +1135,7 @@ static int cpufreq_add_dev(struct device *dev, struct subsys_interface *sif)
 }
 
 static int cpufreq_nominate_new_policy_cpu(struct cpufreq_policy *policy,
-        unsigned int old_cpu, bool frozen)
+        unsigned int old_cpu)
 {
  struct device *cpu_dev;
  int ret;
@@ -1124,10 +1143,6 @@ static int cpufreq_nominate_new_policy_cpu(struct cpufreq_policy *policy,
  /* first sibling now owns the new sysfs dir */
  cpu_dev = get_cpu_device(cpumask_any_but(policy->cpus, old_cpu));
 
- /* Don't touch sysfs files during light-weight tear-down */
- if (frozen)
-  return cpu_dev->id;
-
  sysfs_remove_link(&cpu_dev->kobj, "cpufreq");
  ret = kobject_move(&policy->kobj, &cpu_dev->kobj);
  if (ret) {
@@ -1194,7 +1209,7 @@ static int __cpufreq_remove_dev_prepare(struct device *dev,
   if (!frozen)
    sysfs_remove_link(&dev->kobj, "cpufreq");
  } else if (cpus > 1) {
-  new_cpu = cpufreq_nominate_new_policy_cpu(policy, cpu, frozen);
+  new_cpu = cpufreq_nominate_new_policy_cpu(policy, cpu);
   if (new_cpu >= 0) {
    update_policy_cpu(policy, new_cpu);
 
@@ -1216,8 +1231,6 @@ static int __cpufreq_remove_dev_finish(struct device *dev,
  int ret;
  unsigned long flags;
  struct cpufreq_policy *policy;
- struct kobject *kobj;
- struct completion *cmp;
 
  read_lock_irqsave(&cpufreq_driver_lock, flags);
  policy = per_cpu(cpufreq_cpu_data, cpu);
@@ -1247,22 +1260,8 @@ static int __cpufreq_remove_dev_finish(struct device *dev,
    }
   }
 
-  if (!frozen) {
-   down_read(&policy->rwsem);
-   kobj = &policy->kobj;
-   cmp = &policy->kobj_unregister;
-   up_read(&policy->rwsem);
-   kobject_put(kobj);
-
-   /*
-    * We need to make sure that the underlying kobj is
-    * actually not referenced anymore by anybody before we
-    * proceed with unloading.
-    */
-   pr_debug("waiting for dropping of refcount\n");
-   wait_for_completion(cmp);
-   pr_debug("wait complete\n");
-  }
+  if (!frozen)
+   cpufreq_policy_put_kobj(policy);
 
   /*
    * Perform the ->exit() even during light-weight tear-down,
-- 
1.7.1